Tasks/AzureRmWebAppDeployment/MSDeployUtility.ts

```diff
@@ -0,0 +1,172 @@
+
+export function fileExists(path) {
+  try  {
+    return tl.stats(path).isFile();
+  }
+  catch (e) {
+    if (e.code == 'ENOENT') {
+      return false;
+    }
+    tl.debug("Exception tl.stats (" + path + "): " + e);
+    throw e;
+  }
+}
+
+export function getMSDeployCmdArgs(webAppPackage: string, webAppName: string, publishingProfile,
+                             removeAdditionalFilesFlag: boolean, excludeFilesFromAppDataFlag: boolean, takeAppOfflineFlag: boolean,
+                             virtualApplication: string, setParametersFile: string, additionalArguments: string, isParamFilePresentInPacakge: boolean, isFolderBasedDeployment: boolean) : string {
+
+    var msDeployCmdArgs: string = " -verb:sync";
+
+    var webApplicationDeploymentPath = ( virtualApplication ) ? webAppName + "/" + virtualApplication : webAppName;
+    
+    if(isFolderBasedDeployment) {
+        msDeployCmdArgs += " -source:IisApp='"+ webAppPackage + "'";
+        msDeployCmdArgs += " -dest:iisApp='" + webApplicationDeploymentPath + "',";
+    } else {       
+        msDeployCmdArgs += " -source:package='"+ webAppPackage + "'";
+
+        if(isParamFilePresentInPacakge) {
+            msDeployCmdArgs += " -dest:auto,";           
+        } else {
+            msDeployCmdArgs += " -dest:contentPath='"+ webApplicationDeploymentPath +"',";
+        }
+    }
+
+    msDeployCmdArgs += "ComputerName='https://" + publishingProfile.publishUrl + "/msdeploy.axd?site=" + webAppName + "',";
+    msDeployCmdArgs += "UserName='" + publishingProfile.userName + "',Password='" + publishingProfile.userPWD + "',AuthType='Basic'";
+
+    if( isParamFilePresentInPacakge || setParametersFile != null ){
+        msDeployCmdArgs += " -setParam:name='IIS Web Application Name',value='" + webApplicationDeploymentPath + "'";
+    }
+
+    if (setParametersFile) {
+        
+        msDeployCmdArgs += " -setParamFile=" + setParametersFile;
+    }
+
+    if(!removeAdditionalFilesFlag) {
+        msDeployCmdArgs += " -enableRule:DoNotDeleteRule";
+    }
+
+    if (takeAppOfflineFlag) {
+        msDeployCmdArgs += ' -enableRule:AppOffline';
+    }
+
+    if (excludeFilesFromAppDataFlag) {
+        msDeployCmdArgs += ' -skip:Directory=App_Data';
+    }
+    
+    if (additionalArguments) {
+        msDeployCmdArgs += ' ' + additionalArguments;
+    }
+
+    var userAgent = tl.getVariable("AZURE_HTTP_USER_AGENT");
+    if (userAgent) {
+        msDeployCmdArgs += ' -userAgent:' + userAgent;
+    }
+    tl.debug(tl.loc('ConstructedmsDeploycomamndlinearguments'));
+    return msDeployCmdArgs;
+}
+
+export async function executeMSDeployCmd(msDeployCmdArgs: string, publishingProfile, webAppUri: string) {
+    try {
+        if(webAppUri) {
+            tl.setVariable(webAppUri, publishingProfile.destinationAppUrl);
+        }
+        var msDeployPath = await getMSDeployFullPath();
+            var statusCode = await tl.exec(msDeployPath, msDeployCmdArgs, <any> {failOnStdErr: true});
```

Comment:
Set secure variables for password and username of the website before command is passed to tl.exec. Powershell version to do that is given below,

Write-Host "##vso[task.setvariable variable=$outputVariable;issecret=true;]$envStr"